implementations/rust/message/src/lib.rs

```diff
@@ -0,0 +1,540 @@
+    let mut v: Vec<u8> = vec![0,127,0,0,1];
+    match IpAddr::decode(&v) {
+      Ok((ip4a, w)) => {
+        assert_eq!(ip4a, IpAddr::V4(Ipv4Addr::new(127,0,0,1)));
+      }
+      Err(s) => { println!("{}", s); }
+    }
+  }
+
+  #[test]
+  fn address_codec() {
+    let mut address = Address::UdpAddress(IpAddr::V4(Ipv4Addr::new(127,0,0,1)), 0x8080);
+    let mut v: Vec<u8> = vec![];
+    Address::encode(&mut address, &mut v);
+    assert_eq!(v, vec![2,0,127,0,0,1,0x80,0x80]);
+    let mut v = vec![2,0,127,0,0,1,0x80,0x80];
+    match Address::decode(&mut v) {
+      Ok((address, w)) => {
+        assert_eq!(address, Address::UdpAddress(IpAddr::V4(Ipv4Addr::new(127,0,0,1)), 0x8080));
+      }
+      Err(s) => { println!("{}", s); }
+    }
+    let mut address = Address::LocalAddress( LocalAddress{ length: 4, address:vec![0,1,2,3] } );
+    let mut v: Vec<u8> = vec![];
+    Address::encode(&mut address, &mut v);
+    assert_eq!(v, vec![0,4,0,1,2,3]);
+    let mut v = vec![0,4,0,1,2,3];
+    match Address::decode(&mut v) {
+      Ok((address, w)) => {
+        assert_eq!(address, Address::LocalAddress(LocalAddress{ length: 4, address:vec![0,1,2,3] }));
+      }
+      Err(s) => { println!("{}", s); }
+    }
+  }
+
+  #[test]
+  fn route_codec() {
+    let mut route: Route = Route{ addresses:vec![] };
+    route.addresses.push(Address::UdpAddress(IpAddr::V4(Ipv4Addr::new(127,0,0,1)), 0x8080));
+    route.addresses.push(Address::UdpAddress(IpAddr::V4(Ipv4Addr::new(10,0,1,10)), 0x7070));
+    route.addresses.push(Address::LocalAddress( LocalAddress{ length: 4, address:vec![0,1,2,3] } ));
+    let mut v: Vec<u8> = vec![];
+    Route::encode(&mut route, &mut v);
+    assert_eq!(v, vec![3,2,0,127,0,0,1,0x80,0x80,2,0,10,0,1,10,0x70,0x70,0,4,0,1,2,3]);
+    match Route::decode(&v) {
+      Ok((r, u)) => {
+        assert_eq!(r.addresses.len(), 3);
+        assert_eq!(r.addresses[0], Address::UdpAddress(IpAddr::V4(Ipv4Addr::new(127,0,0,1)), 0x8080));
+        assert_eq!(r.addresses[1], Address::UdpAddress(IpAddr::V4(Ipv4Addr::new(10,0,1,10)), 0x7070));
+        assert_eq!(r.addresses[2], Address::LocalAddress( LocalAddress{ length: 4, address:vec![0,1,2,3] } ));
+        assert_eq!(v.len(), 23);
+      },
+      Err(s) => { panic!(); }
+    }
+    let mut route: Route = Route{ addresses:vec![]};
+    let mut v: Vec<u8> = vec![];
+    Route::encode(&mut route, &mut v);
+    assert_eq!(v.len(), 1);
+    match Route::decode(&v) {
+      Ok((r, v)) => {
+        assert_eq!(r.addresses.len(), 0);
+        assert_eq!(v.len(), 0);
+      },
+      Err(s) => { panic!(); }
+    }
+
+  }
+
+  #[test]
+  #[should_panic]
+  fn u16_codec_too_big() {
+    let mut u: Vec<u8> = vec![];
+    let mut n: u16 = 0xC000;
+    u16::encode(&mut n, &mut u);
+  }
+
+  #[test]
+  fn u16_codec() {
+    let mut u: Vec<u8> = vec![];
+    let mut n: u16 = 0x7f;
+    u16::encode( &mut n, &mut u);
+    assert_eq!(u.len(), 1);
+    assert_eq!(u[0], 0x7f);
```

Comment:
What are these magic byte sequences like `0x7f, 0x80, 0x01`? Could you add a comment for these particular test values

Reply:
Yes comments are in order. They are used when encode variable-length integers. Bit twiddling.